### PR TITLE
Various API helper functions ported to RISC-V, along with callback support and branch printing plugins

### PR DESCRIPTION
**Pull request by jkressel**
I have included the increased size of BRANCH_FSPACE as well as the plugins, since this allows the plugins to run correctly

### CODE REVIEW

**Comment by GuillermoCallaghan** [comment]
Ok, lets start with these steps.

1.  [Increased size of BRANCH_FSPACE to deal with a case where a BNE needed 104 bytes > 96 bytes. Atomic increment u64 for riscv](https://github.com/beehive-lab/mambo/pull/49/commits/515a7dcbaaec0f384213cf6c47ae86feced731f7)
    -  This commit has to be separated into two. One with the increase of `BRANCH_FSPACE` and the other with Atomic increment.
2.  The two "update header" commits need to be squashed together, but the second "update headers" also includes handling  `jal_riscv`, which needs to be in a separate commit. Also includes some code that was uncommented in ```plugins/branch_count.c```  which I think it should not have been commented out in the first place.
3. [Fixed unconditional branch print plugin](Fixed unconditional branch print plugin) this commit should be squashed with the one that introduced the plugin.
4. This one seems ok [Unconditional branch linking on risc v](https://github.com/beehive-lab/mambo/pull/49/commits/94faf566a7b26dc8c35e2f6db7b32f3a7acb52de)
5. [Something is still broken, the plugins mostly work except that when r…](https://github.com/beehive-lab/mambo/pull/49/commits/44f5474d6786753067fd405aec2dccce5bab8594)
   - Ok, if something is not working, don't push it. Also, this includes a lot of things, they need to be in their own commit. For example, once ```cnd_branch_print.c``` is done and working, its commit should be something like "Plugin to print conditional branches"
6.  These commits should not be there. There should be just one that implements the API.
    - [WIP! atomic incrament and RETURN branch decoding still need fixing](https://github.com/beehive-lab/mambo/pull/49/commits/f773dcc0beb6fac3ee7100430ae8214723b4614b)
      - increment has a typo in that commit message. Also I think there is nothing related to _atomic increment_ in this commit.
      - [More API helpers and internals wired up for riscv](https://github.com/beehive-lab/mambo/pull/49/commits/b063f7461a5996b8e33e8b7e5cdb1a025dff449e)
      - [WIP! started wiring up api helper functions on riscv](https://github.com/beehive-lab/mambo/pull/49/commits/3f0bd0794adbf19283025060ae631f474bc03d40)
      - [Started porting api helper functions over to riscv](https://github.com/beehive-lab/mambo/pull/49/commits/c27a19c5e1b3f22a324bcbd050b6eaed4a29739a)
      - [Branch decoder support will now check for return branches on riscv](https://github.com/beehive-lab/mambo/pull/49/commits/34b5d3b78e3779b340eab251b4dd6c31bcd8e7be)
      - [Add beginning of RISCV Branch Decoder - Return instruction decoding i…](https://github.com/beehive-lab/mambo/pull/49/commits/2d44f0726e700c7b648cc7389d2d156da2f1e033)
7. [Elf loader now deals with some unhandled cases for riscv](https://github.com/beehive-lab/mambo/pull/49/commits/86ef62f41cacc1a38daf8d8f3cd7eb945b40df53) These cases are not riscv only, and may need to be handle correctly.

**Comment by jkressel** [comment]
@GuillermoCallaghan @lgeek 
I have fixed the commit situation, so that each feature is a commit

File: plugins/uncnd_branch_print.c [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+#include <stdio.h>
+#include <assert.h>
+#include <locale.h>
+#include <inttypes.h>
+#include "../plugins.h"
+
+int uncnd_branch_print_pre_thread_handler(mambo_context *ctx) {
+  uint64_t *uncnd_br_count = mambo_alloc(ctx, sizeof(uint64_t));
+  *uncnd_br_count = 0;
+  assert(uncnd_br_count != NULL);
```
Comment (GuillermoCallaghan) [inline plugins/uncnd_branch_print.c:29]: The `assert` has to be before you try to write to `uncnd_br_count`.

File: plugins/uncnd_branch_print.c [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  return 0;
+}
+
+// Called when a thread exits, or in all threads when the process exits
+int uncnd_branch_print_post_thread_handler(mambo_context *ctx) {
+  uint64_t *uncnd_br_count = mambo_get_thread_plugin_data(ctx);
+  fprintf(stderr, "%'lu unconditional branches executed in thread %d\n", *uncnd_br_count, mambo_get_thread_id(ctx));
+  mambo_free(ctx, uncnd_br_count);
+  return 0;
+}
```
Comment (GuillermoCallaghan) [inline plugins/uncnd_branch_print.c:40]: `uncnd_branch_print_post_thread_handler` is not incrementing a global counter.
Once a treads finishes you loose the number of unconditional branches that were executed in the thread.

File: plugins/uncnd_branch_print.c [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+*/
+
+#ifdef PLUGINS_NEW
+
+#include <stdio.h>
+#include <assert.h>
+#include <locale.h>
+#include <inttypes.h>
+#include "../plugins.h"
+
```
Comment (GuillermoCallaghan) [inline plugins/uncnd_branch_print.c:25]: It does not have a global counter.

File: plugins/uncnd_branch_print.c [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  mambo_branch_type type = mambo_get_branch_type(ctx);
+  if (!(type & BRANCH_COND || type & BRANCH_NONE)) {
+    emit_counter64_incr(ctx, counter, 1);
+  }
+  return 0;
+}
+
+int uncnd_branch_print_post_inst_handler(mambo_context *ctx) {
+  mambo_branch_type type = mambo_get_branch_type(ctx);
+  if (!(type & BRANCH_COND || type & BRANCH_NONE)) {
```
Comment (GuillermoCallaghan) [inline plugins/uncnd_branch_print.c:53]: Couldn't this be just:
`if (type & BRANCH_COND ) {`

File: plugins/uncnd_branch_print.c [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    fprintf(stderr, "RISCV unconditional branch: read_addr: %p, target: 0x%ld, branch type: %d\n", mambo_get_source_addr(ctx), ctx->thread_data->code_cache_meta[mambo_get_fragment_id(ctx)].branch_taken_addr, ctx->code.inst);
```
Comment (GuillermoCallaghan) [inline plugins/uncnd_branch_print.c:54]: If you want to print the architecture, you need to have a variable that is loaded with architecture when is compiled.
If I execute this plugin on `arm` it will print _"RISCV unconditional branch ...."_ which is not correct.

We also talk about printing the `ctx->thread_data->code_cache_meta[mambo_get_fragment_id(ctx)].branch_taken_addr, ctx->code.inst` in the `post_inst_handler` is not the right thing to do.
In the `pre_inst_handler` you need to have a function call to a function that prints the address of the function.

**Review by GuillermoCallaghan — CHANGES_REQUESTED** [review]

**Comment by lgeek** [comment]
Hi @jkressel. Thanks again for the patches. I think this pull request will be getting very unwieldy if we continue using it for code review, there are many commits, quite a bit of code, and IIRC any time you force push an update, the entire discussion will be hidden so it will be difficult to keep track of what's been done. This is why I was asking for several pull requests more narrow in scope in my initial email. We need to split it up into at least 3 separate pull requests:

* API patches for RISCV: a22387d, 6d7c295, 63a7557, efd3e2f
* Scanner fixes for RISCV: b8e2e9f, ba4fa47
* Plugins: 3fefa17, 1233bf2

Basically create two new branches based on the riscv branch in this repository and cherry-pick the commits I've mentioned. Then open a new pull request for each branch.

My recommendation would be to leave the plugins aside until we get everything else merged and focus on the other patches. Since Guillermo started to review the plugins in this pull request, we can leave it open for them.

**Comment by jkressel** [comment]
@lgeek @GuillermoCallaghan okay I'll create some new pull requests for those commits

File: plugins/uncnd_branch_print.c (hunk 3 above)
Comment (jkressel) [reply plugins/uncnd_branch_print.c:25]: Is a global counter needed?

File: plugins/uncnd_branch_print.c (hunk 4 above)
Comment (jkressel) [reply plugins/uncnd_branch_print.c:53]: Wouldn't that print for conditional branches then?

File: plugins/uncnd_branch_print.c (hunk 5 above)
Comment (jkressel) [reply plugins/uncnd_branch_print.c:54]: Ah okay that makes sense, I'll change that for both

File: plugins/uncnd_branch_print.c (hunk 4 above)
Comment (GuillermoCallaghan) [reply plugins/uncnd_branch_print.c:53]: Sorry, that should have been this:
`if (type & BRANCH_DIRECT) {`

File: plugins/uncnd_branch_print.c (hunk 3 above)
Comment (GuillermoCallaghan) [reply plugins/uncnd_branch_print.c:25]: You need a global counter to keep the numbers of the branches of each thread so at the end you can print the total number of branches for the application.
Have a look at how the global counter is used in [`branch_count.c`](https://github.com/jkressel/mambo/blob/riscv-plugins/plugins/branch_count.c#L29) and how is handled in the [`post_inst_handler`](https://github.com/jkressel/mambo/blob/riscv-plugins/plugins/branch_count.c#L72).

File: plugins/uncnd_branch_print.c (hunk 4 above)
Comment (jkressel) [reply plugins/uncnd_branch_print.c:53]: Doesn't a BRANCH_DIRECT just refer to a branch to an immediate address?

Comment (GuillermoCallaghan) [reply plugins/uncnd_branch_print.c:53]: You are **right**. I got confused.

File: plugins/uncnd_branch_print.c (hunk 5 above)
Comment (GuillermoCallaghan) [reply plugins/uncnd_branch_print.c:54]: No, you need to insert a safe function call on the pre instruction, that should be a call to a function that takes two arguments, one is the address where the branch instruction is and the other is what type of branch it is and prints these two things.

Comment (jkressel) [reply plugins/uncnd_branch_print.c:54]: Ah I see, that way every time the branch is executed, the function is also called

Comment (GuillermoCallaghan) [reply plugins/uncnd_branch_print.c:54]: Correct!
